File: libcirc/innerprod.c
#include <stdlib.h>
#include <math.h>

#include "stabilizer/stabilizer.h"
#include "mpi.h"

/***************** Prototypes *******************/
// linked from stateprep.c
struct StabilizerState* prepH(int i, int t);
struct StabilizerState* prepL(int i, int t, struct BitMatrix* L);

double sampledProjector(struct Projector *P, struct BitMatrix *L, int exact, double norm, int samples);
double singleProjectorSample(struct Projector *P, struct BitMatrix *L, int exact);
Complex exactProjectorWork(int i, struct Projector *P, struct BitMatrix *L, int exact);

// sort compare function
int cmpfunc(const void* a, const void* b) {
   return *(double*)a - *(double*)b;
}

/******************* Code **********************/
// Master function. Calculate median of several means. 
double multiSampledProjector(struct Projector *P, struct BitMatrix *L, int exact, double norm, int samples, int bins) {
    if (bins == 1) return sampledProjector(P, L, exact, norm, samples);
   
    double* binVals = malloc(bins * sizeof(double));
    for (int i = 0; i < bins; i++) {
        binVals[i] = sampledProjector(P, L, exact, norm, samples);
    }

    qsort(binVals, bins, sizeof(double), cmpfunc);

    double out = 0;
    if (bins % 2 == 1) out = binVals[(bins-1)/2];
    else {
        out = (binVals[bins/2] + binVals[bins/2 - 1])/2;
    }
    
    free(binVals);
    return out;
}


// Master function. Calculate mean of several samples.
double sampledProjector(struct Projector* P, struct BitMatrix* L, int exact, double norm, int samples) {
    // empty projector
    if (P->Nstabs == 0) return pow(norm, 2);

    int t = P->Nqubits;

    // clifford circuit
    if (t == 0) {
        double sum = 1; // include identity
        for (int i = 0; i < P->Nstabs; i++) {
            double ph = (double)BitVectorGet(P->phaseSign, i)*2;
            ph += (double)BitVectorGet(P->phaseComplex, i);
            if (ph == 0) sum += 1;
            if (ph == 2) sum -= 1;
        } 

        return pow(norm, 2) * sum/(1 + (double)P->Nstabs);
    }

    // parallel code
    int world_size;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);

    double total = 0;
    for (int dest = 1; dest < world_size; dest++) {
        sendInt(2, dest); // command
        sendProjector(P, dest);
        sendInt(samples, dest);
    }    

    for (int i = 0; i < samples; i += world_size) {
        total += singleProjectorSample(P, L, exact);
    }

    for (int src = 1; src < world_size; src++) {
        total += recvDouble(src);
    }

    return total/samples;
}


// Evaluate 2^t * || <theta| P |H^t> ||^2 for a random theta.
double singleProjectorSample(struct Projector *P, struct BitMatrix *L, int exact) {
    int t = P->Nqubits;

    int k;
    if (!exact) k = L->rows;

    // Sample random stabilizer state
	struct StabilizerState *theta = randomStabilizerState(t);

    // project state onto P
    double projfactor = 1;

    for (int i = 0; i < P->Nstabs; i++) {
        int m = BitVectorGet(P->phaseSign, i)*2;
        m += BitVectorGet(P->phaseComplex, i);
        struct BitVector* zeta = BitMatrixGetRow(P->zs, i);
        struct BitVector* xi = BitMatrixGetRow(P->xs, i);

        double res = measurePauli(theta, m, zeta, xi);
        projfactor *= res;

        BitVectorFree(zeta);
        BitVectorFree(xi);

        if (res == 0) {
            freeStabilizerState(theta);
            return 0;
        }
    } 


    struct StabilizerState *phi;
    Complex total = {0,0};

    if (exact) {
        int size = pow(2, ceil((double)t / 2));
        for (int i = 0; i < size; i+=1) {
            phi = prepH(i, t);
            Complex innerProd = innerProduct(theta, phi);
            total = ComplexAdd(total, innerProd);
            freeStabilizerState(phi);
        }
    } else {
        int size = pow(2, k);
        for (int i = 0; i < size; i+=1) {
            phi = prepL(i, t, L);
            Complex innerProd = innerProduct(theta, phi);
            total = ComplexAdd(total, innerProd);
            freeStabilizerState(phi);
        }
    }
    
    freeStabilizerState(theta);

    double out = pow(2, t) * ComplexMagSquare(ComplexMulReal(total, projfactor));
    return out;
}


// Master function. Calculate projector exactly. 
double exactProjector(struct Projector* P, struct BitMatrix* L, int exact, double norm) {
    // empty projector
    if (P->Nstabs == 0) return pow(norm, 2);

    int t = P->Nqubits;
    int k;
    if (!exact) k = L->rows;

    // clifford circuit
    if (t == 0) {
        double sum = 1; // include identity
        for (int i = 0; i < P->Nstabs; i++) {
            double ph = (double)BitVectorGet(P->phaseSign, i)*2;
            ph += (double)BitVectorGet(P->phaseComplex, i);
            if (ph == 0) sum += 1;
            if (ph == 2) sum -= 1;
        } 

        return sum/(1 + (double)P->Nstabs);
    }
    
    // parallel code
    int world_size;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);

    int size;
    if (exact) size = ceil((double)t / 2);
    else size = k;

    int kRange = pow(2,size - 1) * (pow(2, size) +1);  // chi * (chi + 1)/2

    Complex total = {0,0};
    Complex part;
    for (int dest = 1; dest < world_size; dest++) {
        sendInt(2, dest); // command
        sendProjector(P, dest);
        sendInt(kRange, dest);
    }    

    for (int l = 0; l < kRange; l += world_size) {
        part = exactProjectorWork(l, P, L, exact);
        total = ComplexAdd(total, part);
    }

    for (int src = 1; src < world_size; src++) {
        part = recvComplex(src);
        total = ComplexAdd(total, part);
    }


    return ComplexMag(total);
}


// Work function for exactProjector.
Complex exactProjectorWork(int l, struct Projector* P, struct BitMatrix* L, int exact) {
    int t = P->Nqubits;

    int chi;
    if (exact) chi = pow(2, ceil((double)t / 2));
    else chi = pow(2, L->rows);

    int i = 0;
    while (l >= chi - i) {
        l -= chi - i;
        i += 1;
    }
    int j = l + i;

    struct StabilizerState* theta;
    if (exact) theta = prepH(i, t);
    else theta = prepL(i, t, L);

    // Project theta
    double projfactor = 1;

    for (int r = 0; r < P->Nstabs; r++) {
        int m = BitVectorGet(P->phaseSign, r)*2;
        m += BitVectorGet(P->phaseComplex, r);
        struct BitVector* zeta = BitMatrixGetRow(P->zs, r);
        struct BitVector* xi = BitMatrixGetRow(P->xs, r);
       
        double res = measurePauli(theta, m, zeta, xi);
        projfactor *= res;
        
        BitVectorFree(zeta);
        BitVectorFree(xi);

        if (res == 0) {
            freeStabilizerState(theta);
            Complex out = {0,0};
            return out;
        }
    } 
    // Evaluate other components
    struct StabilizerState* phi;

    // Diagonal term
    if (exact) phi = prepH(j, t);
    else phi = prepL(j, t, L);

    Complex innerProd = innerProduct(theta, phi);

    freeStabilizerState(phi);
    freeStabilizerState(theta);
   
    if (i == j) {
        return ComplexMulReal(innerProd, projfactor); 
    } else {
        innerProd.re *= 2*projfactor;
        innerProd.im = 0;
        return innerProd; 
    }
}

File: libcirc/innerprod.py
#
# file: innerprod.py
# Methods for evaluating inner products of P|H^t> where
# P is a given projector.
#

import numpy as np
from multiprocessing import Pool, cpu_count
from libcirc.stabilizer.stabilizer import StabilizerState
from libcirc.stateprep import prepH, prepL


# Median of means calculation can be done via Chebychev and Chernoff bounds.
# http://www.cs.utexas.edu/~ecprice/courses/randomized/notes/lec5.pdf
# If sampledProjector has error worse than e with probability p,
# then multiSampledProjector has error worse than e with probability
# less than delta = exp(-2m(1/2 - p)^2) where m is the number of bins.
#
# Since L is proportional to 1/p, best total number of samples m*L is minimized
# at the minimum of m/p proportional to p^(-1) * (1/2 - p)^(-2) which is p = 1/6.
# Thus for best results pick L = 6 * (2^t - 1)/(2^t + 1) * (1/e^2)
# and m = 4.5 log(1 / delta). This is only better than mean-only sampling
# for failure probabilities smaller than 0.0076 which is the solution to
# the equation 1/p = 27 ln (1/p).
#
# In the large t limit, can achieve e = 0.1 with 99.9% chance with 600 samples,
# and 32 bins. This would need 100 000 samples for mean-only mode, over 5x more.
def multiSampledProjector(P, L, norm, bins=32, samples=600, procs=1):
    means = []
    for i in range(bins):
        means.append(sampledProjector(P, L, norm, samples=samples, procs=procs))
    return np.median(means)


# If number of samples is L, then this distribution has mean || P |H^t> ||^2
# and standard deviation sqrt((2^t-1)/(2^t+1)) * || P |H^t> ||^2 * (1/sqrt(L))
#
# On its own, this method guarantees output error e * || P |H^t> ||^2
# with probability (1 - p) if L = (2^t - 1)/(2^t + 1) * (1/p) * (1/e^2)
#
# In the large t limit, can achieve e = 0.1 with 95% chance with 2000 samples.
def sampledProjector(P, L, norm, samples=2000, procs=1):
    (phases, xs, zs) = P

    # empty projector
    if len(phases) == 0:
        return np.abs(norm)**2

    # clifford circuit
    if len(xs[0]) == 0:
        lookup = {0: 1, 2: -1}
        gens = [1]  # include identity
        for phase in phases: gens.append(lookup[phase])

        # calculate sum of all permutations of generators
        return sum(gens)/len(gens) * np.abs(norm)**2

    if procs is None:
        try:
            procs = cpu_count()
        except NotImplementedError:
            procs = 1

    seeds = np.random.random_integers(0, 2**32-1, samples)
    queries = [(P, L, seed) for seed in seeds]
    if procs > 1:
        pool = Pool(procs)
        out = sum(pool.map(singleProjectorSample, queries))/samples
        pool.close()
        return out
    else:
        return sum(map(singleProjectorSample, queries))/samples


# Evaluate 2^t * || <theta| P |H^t> ||^2 for a random theta.
# This distribution has mean || P |H^t> ||^2
# and standard deviation sqrt((2^t-1)/(2^t+1)) * || P |H^t> ||^2
def singleProjectorSample(args):
    (P, L, seed) = args  # unpack arguments (easier for parallel code)
    (phases, xs, zs) = P
    t = len(xs[0])

    # init seed
    np.random.seed(seed)

    # sample random theta
    theta = StabilizerState.randomStabilizerState(t)

    # project random state to P
    projfactor = 1
    for g in range(len(phases)):
        res = theta.measurePauli(phases[g], zs[g], xs[g])
        projfactor *= res

        if res == 0: return 0  # theta annihilated by P

    total = 0
    if L is None:  # exact decomposition
        size = int(np.ceil(t/2))
        for i in range(0, 2**size):
            phi = prepH(i, t)
            total += StabilizerState.innerProduct(theta, phi)

    else:  # approximate decomposition
        size = len(L)
        for i in range(0, 2**size):
            phi = prepL(i, t, L)
            total += StabilizerState.innerProduct(theta, phi)

    return 2**t * np.abs(projfactor*total)**2


# calculate projector exactly
def exactProjector(P, L, norm, procs=1):
    (phases, xs, zs) = P

    # empty projector
    if len(phases) == 0:
        return np.abs(norm)**2

    t = len(xs[0])

    # clifford circuit
    if t == 0:
        lookup = {0: 1, 2: -1}
        generators = [1]  # include identity
        for phase in phases: generators.append(lookup[phase])

        # calculate sum of all permutations of generators
        return sum(generators)/len(generators)

    if L is None: size = int(np.ceil(t/2))
    else: size = len(L)

    if procs is None:
        try:
            procs = cpu_count()
        except NotImplementedError:
            procs = 1

    queries = [(P, L, l) for l in range(0, 2**(size-1) * (2**size + 1))]
    if procs > 1:
        pool = Pool(procs)
        total = sum(pool.map(exactProjectorWork, queries))
        pool.close()
    else:
        total = sum(map(exactProjectorWork, queries))
    return np.abs(total)


def exactProjectorWork(args):
    (P, L, l) = args
    (phases, xs, zs) = P

    t = len(xs[0])
    if L is None: size = int(np.ceil(t/2))
    else: size = len(L)

    chi = 2**size

    i = 0
    while l >= chi - i:
        l -= chi - i
        i += 1
    j = l + i

    if L is None: theta = prepH(i, t)
    else: theta = prepL(i, t, L)

    projfactor = 1
    for g in range(len(phases)):
        res, status = theta.measurePauli(phases[g], zs[g], xs[g], give_status=True)

        projfactor *= res

        if res == 0: return 0  # theta annihilated by P

    if L is None: phi = prepH(j, t)
    else: phi = prepL(j, t, L)

    inner = StabilizerState.innerProduct(theta, phi)
    if i == j:
        return inner * projfactor
    else:
        return 2 * np.real(inner) * projfactor
